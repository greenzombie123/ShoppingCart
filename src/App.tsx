import { useState } from 'react'

function App({title}:{title:string}) {

  return (
    <>
    
    </>
  )
}

export default App
