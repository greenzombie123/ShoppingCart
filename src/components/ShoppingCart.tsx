import {
  Form,
  Link,
  Outlet,
  useFetcher,
  useLoaderData,
} from "react-router-dom";
import { Cart, CartItem, Product } from "../products";
import style from "./ShoppingCart.module.css";
import { StarContainer } from "./StorePage";
import { useState } from "react";
import CartPopUp from "./CartPopUp";
import useCartItems from "../custom_hooks/useCartItems";

export const ViewedItemsContainer = () => {
  const viewedItems = useLoaderData<Product[]>();

  return (
    <div className={style.viewedList}>
      <p className={style.heading}>Viewed Items</p>
      {viewedItems.map((product) => (
        <ViewedItem key={product.id} product={product} />
      ))}
    </div>
  );
};

type ViewedItemProp = {
  product: Product;
};

const ViewedItem = ({ product }: ViewedItemProp) => {
  return (
    <div className={style.viewedItem}>
      <Link to={`product/${product.id}`}>
        <div className={style.imgContainer}>
          <img src={product.styles[0].picture} alt={product.name} />
        </div>
      </Link>
      <div className={style.infoContainer}>
        <div className={style.name}>{product.name}</div>
        <div className={style.stars}>
          <StarContainer stars={product.stars} style={style} />
          <div className={style.ratings}>({product.ratings})</div>
        </div>
        <div className={style.price}>${product.price}</div>
        <button className={style.addButton}>Add to Cart</button>
      </div>
    </div>
  );
};

const QuantityCounter = ({
  cartItem,
  onIncreaseButtonClick,
  onDecreaseButtonClick,
}: {
  cartItem: CartItem;
  onIncreaseButtonClick: (id: number) => void;
  onDecreaseButtonClick: (id: number) => void;
}) => {
  // console.log(cartItem.quantity)

  return (
    <div className={style.quantityCounter}>
      <button
        type="button"
        id="increaseButton"
        aria-label="Increase quantity"
        className={style.increaseButton}
        onClick={() => onIncreaseButtonClick(cartItem.id)}
      >
        &#43;
      </button>
      <output
        htmlFor="increaseButton decreaseButton"
        name="quantity"
        className={style.counter}
      >
        {cartItem.quantity}
      </output>
      <button
        type="button"
        id="decreaseButton"
        aria-label="Decrease quantity"
        className={style.decreaseButton}
        onClick={() => onDecreaseButtonClick(cartItem.id)}
      >
        &#8722;
      </button>
    </div>
  );
};

const Item = ({
  cartItem,
  onIncreaseButtonClick,
  onDecreaseButtonClick,
  setPopUp,
}: {
  cartItem: CartItem;
  onIncreaseButtonClick: (id: number) => void;
  onDecreaseButtonClick: (id: number) => void;
  setPopUp: React.Dispatch<React.SetStateAction<CartItem | null>>;
}) => {
  const handleRemoveButtonClick = () => {
    setPopUp(cartItem);
  };

  return (
    <div className={style.item}>
      <Link to={`/product/${cartItem.id}`}>
        <div className={style.imgContainer}>
          <img src={cartItem.picture} alt={cartItem.name} />
        </div>
      </Link>
      <div className={style.itemInfo_left}>
        <p className={style.name}>{cartItem.name}</p>
        {cartItem.style && <p className={style.style}>{cartItem.style}</p>}
        <button
          type="button"
          aria-label="remove Cartitem Button"
          className={style.removeButton}
          onClick={handleRemoveButtonClick}
        >
          Remove
        </button>
      </div>
      <div className={style.itemInfo_right}>
        <QuantityCounter
          cartItem={cartItem}
          onDecreaseButtonClick={onDecreaseButtonClick}
          onIncreaseButtonClick={onIncreaseButtonClick}
        />

        <p className={style.price}>
          {"$" + (cartItem.price * cartItem.quantity).toFixed(2)}
        </p>
      </div>
    </div>
  );
};

const ShoppingCart = () => {
  const cart = useLoaderData<Cart>();
  const { cartItems, handleQuantityDecrease, handleQuantityIncrease } =
    useCartItems(cart);
  const [popUpData, setPopUpData] = useState<CartItem | null>(null);

  console.log(cart)

  return (
    <div className={style.shoppingCart}>
      <div className={style.cart}>
        {cart.map((cartItem, index) => (
          <Item
            cartItem={{ ...cartItem, quantity: cartItems[index].quantity }}
            key={cartItem.id}
            onDecreaseButtonClick={handleQuantityDecrease}
            onIncreaseButtonClick={handleQuantityIncrease}
            setPopUp={setPopUpData}
          />
        ))}
      </div>
      <div className={style.rightSide}>
        <Form className={style.priceContainer}>
          <div className={style.topPrice}>
            <p>Subtotal</p>
            <p>$1200</p>
          </div>
          <div className={style.bottomPrice}>
            <p>Total Incl. Tax</p>
            <p>$1222</p>
          </div>
          <button className={style.checkoutButton}>Checkout</button>
        </Form>

        <Outlet />
      </div>

      <CartPopUp popUpData={popUpData} setPopUp={setPopUpData} />
    </div>
  );
};

export default ShoppingCart;
