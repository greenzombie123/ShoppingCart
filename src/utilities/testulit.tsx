import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ReactNode } from "react";
import {
  ActionFunction,
  createMemoryRouter,
  LoaderFunction,
  RouteObject,
  RouterProvider,
} from "react-router-dom";
import { Cart, CartItem, Product } from "../products";
import { vi } from "vitest";

export type RouteObjectProps = {
  element: ReactNode;
  path?: string;
  loader?: LoaderFunction;
  action?: ActionFunction;
  children?: RouteObject[];
};

export const renderWithRouter = ({
  element,
  path = "/",
  loader,
  action,
  children = [],
}: RouteObjectProps) => {
  const router = createMemoryRouter(
    [{ element, path, loader, action, children }],
    {
      initialEntries: [path],
    }
  );

  return {
    user: userEvent.setup(),
    ...render(<RouterProvider router={router} />),
  };
};

export const mockCart: Cart = [
  {
    cartItemId: "asd1",
    name: "LBJ Boom Box",
    id: 12,
    price: 59.99,
    style: "Red",
    picture: "/images/redjbl-boombox.jpeg",
    quantity: 2,
    product: {} as Product,
  },
  {
    cartItemId: "asd2",
    name: "Maggie Lo Blouse",
    id: 13,
    price: 39.99,
    style: "Black",
    picture: "/images/brownblouse.jpg",
    quantity: 1,
    product: {} as Product,
  },
  {
    cartItemId: "asd3",
    name: "Mens Cotton Jacket",
    id: 16,
    price: 55.99,
    style: undefined,
    picture: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
    quantity: 4,
    product: {} as Product,
  },
];

export const mockOneProduct: Product = {
  name: "Jupopo AirFlex Running Shoes",
  id: "7",
  category: "Men's Clothing",
  price: 89.99,
  ratings: 198,
  stars: 4,
  likes: 431,
  styles: [
    {
      description: "",
      picture: "/images/ID3692_HM5.avif",
      isCurrentStyle: true,
    },
  ],
};

export const mockOneCartItem: CartItem = {
  name: "Jupopo AirFlex Running Shoes",
  price: 89.99,
  id: "123",
  quantity: 2,
  style: "",
  picture: "/images/ID3692_HM5.avif",
  product: mockOneProduct,
};

export const mockProducts: Product[] = [
  {
    name: "Jupopo AirFlex Running Shoes",
    id: 7,
    category: "Men's Clothing",
    price: 89.99,
    ratings: 198,
    stars: 4,
    likes: 431,
    styles: [
      {
        description: "",
        picture: "/images/ID3692_HM5.avif",
        isCurrentStyle: true,
      },
    ],
  },
  {
    name: "Velmora Elegance Handbag",
    id: 8,
    category: "Women's Clothing",
    price: 120,
    ratings: 564,
    stars: 5,
    likes: 1201,
    styles: [
      {
        description: "Black",
        picture: "/images/handbagblack.webp",
        isCurrentStyle: true,
      },
      {
        description: "Beige",
        picture: "/images/1_536aef2b-2ba4-4bb5-98ec-6d6ac443cb43.webp",
        isCurrentStyle: false,
      },
    ],
  },
  {
    name: "Orvella Silver Twist Earrings",
    id: 9,
    category: "Jewelry",
    price: 59.99,
    ratings: 310,
    stars: 4,
    likes: 720,
    styles: [
      {
        description: "",
        picture: "/images/twistearrings.webp",
        isCurrentStyle: true,
      },
    ],
  },
  {
    name: "AxionBeat Wireless Headphones",
    id: 10,
    category: "Electronics",
    price: 199.99,
    ratings: 785,
    stars: 5,
    likes: 1532,
    styles: [
      {
        description: "Blue",
        picture: "/images/headlessblue.webp",
        isCurrentStyle: true,
      },
      {
        description: "White",
        picture: "/images/headlesswhite.webp",
        isCurrentStyle: false,
      },
    ],
  },
];

// export const mockModule = async <U, T extends Partial<U>>(
//   path: string,
//   newProps: T
// ): Promise<void> => {

//   vi.mock(path, async () => {
//     const originalModule = (await vi.importActual(path)) as U;

//     return {
//       ...originalModule,
//       ...newProps,
//     };
//   });
// };

export const mockGetOneViewedItem: () => Product[] = () => [
  {
    name: "Jupopo AirFlex Running Shoes",
    id: 7,
    category: "Men's Clothing",
    price: 89.99,
    ratings: 198,
    stars: 4,
    likes: 431,
    styles: [
      {
        description: "",
        picture: "/images/ID3692_HM5.avif",
        isCurrentStyle: true,
      },
    ],
  },
];

export const mockGetEmptyCart: () => [] = () => [];

export const mockAddOneItemToCart: () => { productInfo: CartItem } = () => ({
  productInfo: {
    cartItemId: "111",
    name: "Red Boom Box",
    id: 1,
    price: 59.99,
    style: "Red",
    picture: "/images/redjbl-boombox.jpeg",
    quantity: 3,
    product: {} as Product,
  },
});

export const setUpMockShowModal = () => {
  const mockShowModal = vi.fn();

  HTMLDialogElement.prototype.showModal = mockShowModal;

  return mockShowModal;
};
