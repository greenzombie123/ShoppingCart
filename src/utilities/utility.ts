export const changeToPrice = (price: number) => {
  const string = (price / 100).toFixed(2).toString();
  if (string.length >= 8 && string.length <= 9) {
    const backPart = string.slice(-6);
    const frontPart = string.slice(0, -6);
    return "$" + frontPart + "," + backPart;
  } else return "$" + string;
};
