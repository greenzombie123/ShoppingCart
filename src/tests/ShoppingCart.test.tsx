import { describe, expect, it, vi } from "vitest";
import ShoppingCart from "../components/ShoppingCart";
import { mockCart, renderWithRouter } from "../utilities/testulit";
import { findByRole, screen, waitFor, within } from "@testing-library/dom";
import {
  createMemoryRouter,
  RouteObject,
  RouteProps,
  RouterProvider,
} from "react-router-dom";
import App from "../App";
import userEvent from "@testing-library/user-event";
import { act, render } from "@testing-library/react";

describe("ShoppingCart", () => {
  it("renders cart item", async () => {
    const route = {
      element: <ShoppingCart />,
      path: "/",
      loader: () => mockCart,
    };

    renderWithRouter(route);

    const images = (await screen.findAllByRole("img")) as HTMLImageElement[];
    const cartItem1Name = await screen.findByText("LBJ Boom Box");
    const cartItem1Style = await screen.findByText("Red");
    const cartItem1Quantity = await screen.findByText("2");
    const cartItem1Price = await screen.findByText("$119.98");
    const imgSrc = /\/images\/redjbl-boombox.jpe/;

    expect(cartItem1Name).toBeInTheDocument();
    expect(cartItem1Style).toBeInTheDocument();
    expect(cartItem1Quantity).toBeInTheDocument();
    expect(cartItem1Price).toBeInTheDocument();
    if (images[0]) expect(imgSrc.test(images[0].src)).toBe(true);
  });

  it("is rendered when the cart icon is clicked", async () => {
    const route: RouteObject[] = [
      {
        element: <App />,
        path: "/",
        loader: () => mockCart,
        children: [
          {
            element: <ShoppingCart />,
            path: "/mycart",
            loader: () => mockCart,
          },
          {
            element: <div>Mock</div>,
            index: true,
          },
        ],
      },
    ];
    const user = userEvent.setup();
    const router = createMemoryRouter(route);
    render(<RouterProvider router={router} />);

    const cartIcon = await screen.findByRole("link", { name: "cartIcon" });
    await user.click(cartIcon);

    const cartItem1Name1 = await screen.findByText("LBJ Boom Box");
    expect(cartItem1Name1).toBeInTheDocument();

    const cartItem1Name2 = await screen.findByText("Maggie Lo Blouse");
    expect(cartItem1Name2).toBeInTheDocument();
  });

  it("renders a quantity counter that shows the quantity of a cart item", async () => {
    const route = {
      element: <ShoppingCart />,
      path: "/",
      loader: () => [
        {
          name: "LBJ Boom Box",
          id: 12,
          price: 59.99,
          style: "Red",
          picture: "/images/redjbl-boombox.jpeg",
          quantity: 10,
        },
      ],
    };

    const { findByRole } = renderWithRouter(route);

    const increaseButton = (await findByRole("button", {
      name: "Increase quantity",
    })) as HTMLButtonElement;

    const decreaseButton = (await findByRole("button", {
      name: "Decrease quantity",
    })) as HTMLButtonElement;

    const quantityCounter = await findByRole("status");

    expect(quantityCounter.textContent).toBe("10");
    expect(increaseButton).toBeInTheDocument();
    expect(decreaseButton).toBeInTheDocument();
  });

  it("changes quantity of cart items to 2 when quantity counter is clicked", async () => {
    const route = {
      element: <ShoppingCart />,
      path: "/",
      loader: () => [
        {
          name: "LBJ Boom Box",
          id: 12,
          price: 59.99,
          style: "Red",
          picture: "/images/redjbl-boombox.jpeg",
          quantity: 2,
        },
      ],
    };

    const { user, findByRole } = renderWithRouter(route);
    const increaseButton = (await findByRole("button", {
      name: "Increase quantity",
    })) as HTMLButtonElement;
    const quantityCounter = await findByRole("status");

    await waitFor(async () => {
      await user.click(increaseButton);
    });

    expect(quantityCounter.textContent).toBe("3");

    await waitFor(async () => {
      await user.click(increaseButton);
    });

    expect(quantityCounter.textContent).toBe("4");
  });

  it("renders a popup confirming if you want to remove cart when remove button is clicked", async () => {
    HTMLDialogElement.prototype.showModal = vi.fn();
    HTMLDialogElement.prototype.close = vi.fn();

    const route = {
      element: <ShoppingCart />,
      path: "/",
      loader: () => [
        {
          name: "LBJ Boom Box",
          id: 12,
          price: 59.99,
          style: "Red",
          picture: "/images/redjbl-boombox.jpeg",
          quantity: 2,
        },
      ],
    };

    const { user, findByRole } = renderWithRouter(route);

    const removeButton = (await findByRole("button", {
      name: "remove Cartitem Button",
    })) as HTMLButtonElement;

    await user.click(removeButton);

    expect(HTMLDialogElement.prototype.showModal).toBeCalled();

    vi.resetAllMocks();
  });

  it.skip("removes the cart item from server when ok button of cart popup is clicked", async () => {
    const mockAction = vi.fn();

    const route = {
      element: <ShoppingCart />,
      path: "/mycart",
      loader: () => [
        {
          name: "LBJ Boom Box",
          id: 12,
          price: 59.99,
          style: "Red",
          picture: "/images/redjbl-boombox.jpeg",
          quantity: 2,
        },
      ],
      action: mockAction,
    };

    // const { user, findByRole } = renderWithRouter(route);

    const user = userEvent.setup();

    const { findByRole } = render(
      <RouterProvider
        router={createMemoryRouter([route], { initialEntries: ["/mycart"] })}
      />
    );

    const removeButton = (await findByRole("button", {
      name: "remove Cartitem Button",
    })) as HTMLButtonElement;

    await waitFor(async () => {
      await user.click(removeButton);
    });

    const dialog = await findByRole("dialog", { hidden: true });
    const okButton = dialog.querySelector("button") as HTMLButtonElement;

    await user.click(okButton);

    // expect(mockAction).toBeCalled()
  });

  it.skip("renders the viewed items container", () => {
    const ViewedItemsRoute = {
      element: <ShoppingCart />,
      path: "/",
      loader: () => mockCart,
    };

    const { user, findByRole } = renderWithRouter(route);
  });
});
