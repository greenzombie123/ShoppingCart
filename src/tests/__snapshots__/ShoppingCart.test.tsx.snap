// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingCart > calls addToCart 1`] = `
<div>
  <div
    class="_shoppingCart_fa5548"
  >
    <div
      class="_cart_fa5548"
    >
      <div
        class="_item_fa5548"
      >
        <a
          data-discover="true"
          href="/product/12"
        >
          <div
            class="_imgContainer_fa5548"
          >
            <img
              alt="LBJ Boom Box"
              src="/images/redjbl-boombox.jpeg"
            />
          </div>
        </a>
        <div
          class="_itemInfo_left_fa5548"
        >
          <p
            class="_name_fa5548"
          >
            LBJ Boom Box
          </p>
          <p
            class="_style_fa5548"
          >
            Red
          </p>
          <button
            aria-label="remove Cartitem Button"
            class="_removeButton_fa5548"
            type="button"
          >
            Remove
          </button>
        </div>
        <div
          class="_itemInfo_right_fa5548"
        >
          <div
            class="_quantityCounter_fa5548"
          >
            <button
              aria-label="Increase quantity"
              class="_increaseButton_fa5548"
              id="increaseButton"
              type="button"
            >
              +
            </button>
            <output
              class="_counter_fa5548"
              for="increaseButton decreaseButton"
              name="quantity"
            >
              2
            </output>
            <button
              aria-label="Decrease quantity"
              class="_decreaseButton_fa5548"
              id="decreaseButton"
              type="button"
            >
              −
            </button>
          </div>
          <p
            class="_price_fa5548"
          >
            $119.98
          </p>
        </div>
      </div>
      <div
        class="_item_fa5548"
      >
        <a
          data-discover="true"
          href="/product/13"
        >
          <div
            class="_imgContainer_fa5548"
          >
            <img
              alt="Maggie Lo Blouse"
              src="/images/brownblouse.jpg"
            />
          </div>
        </a>
        <div
          class="_itemInfo_left_fa5548"
        >
          <p
            class="_name_fa5548"
          >
            Maggie Lo Blouse
          </p>
          <p
            class="_style_fa5548"
          >
            Black
          </p>
          <button
            aria-label="remove Cartitem Button"
            class="_removeButton_fa5548"
            type="button"
          >
            Remove
          </button>
        </div>
        <div
          class="_itemInfo_right_fa5548"
        >
          <div
            class="_quantityCounter_fa5548"
          >
            <button
              aria-label="Increase quantity"
              class="_increaseButton_fa5548"
              id="increaseButton"
              type="button"
            >
              +
            </button>
            <output
              class="_counter_fa5548"
              for="increaseButton decreaseButton"
              name="quantity"
            >
              1
            </output>
            <button
              aria-label="Decrease quantity"
              class="_decreaseButton_fa5548"
              id="decreaseButton"
              type="button"
            >
              −
            </button>
          </div>
          <p
            class="_price_fa5548"
          >
            $39.99
          </p>
        </div>
      </div>
      <div
        class="_item_fa5548"
      >
        <a
          data-discover="true"
          href="/product/16"
        >
          <div
            class="_imgContainer_fa5548"
          >
            <img
              alt="Mens Cotton Jacket"
              src="https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg"
            />
          </div>
        </a>
        <div
          class="_itemInfo_left_fa5548"
        >
          <p
            class="_name_fa5548"
          >
            Mens Cotton Jacket
          </p>
          <button
            aria-label="remove Cartitem Button"
            class="_removeButton_fa5548"
            type="button"
          >
            Remove
          </button>
        </div>
        <div
          class="_itemInfo_right_fa5548"
        >
          <div
            class="_quantityCounter_fa5548"
          >
            <button
              aria-label="Increase quantity"
              class="_increaseButton_fa5548"
              id="increaseButton"
              type="button"
            >
              +
            </button>
            <output
              class="_counter_fa5548"
              for="increaseButton decreaseButton"
              name="quantity"
            >
              4
            </output>
            <button
              aria-label="Decrease quantity"
              class="_decreaseButton_fa5548"
              id="decreaseButton"
              type="button"
            >
              −
            </button>
          </div>
          <p
            class="_price_fa5548"
          >
            $223.96
          </p>
        </div>
      </div>
    </div>
    <div
      class="_rightSide_fa5548"
    >
      <form
        action="/mycart"
        class="_priceContainer_fa5548"
        data-discover="true"
        method="get"
      >
        <div
          class="_topPrice_fa5548"
        >
          <p>
            Subtotal
          </p>
          <p>
            $1200
          </p>
        </div>
        <div
          class="_bottomPrice_fa5548"
        >
          <p>
            Total Incl. Tax
          </p>
          <p>
            $1222
          </p>
        </div>
        <button
          class="_checkoutButton_fa5548"
        >
          Checkout
        </button>
      </form>
      <div
        class="_viewedList_fa5548"
      >
        <p
          class="_heading_fa5548"
        >
          Viewed Items
        </p>
        <div
          class="_viewedItem_fa5548"
        >
          <a
            data-discover="true"
            href="/product/7"
          >
            <div
              class="_imgContainer_fa5548"
            >
              <img
                alt="Jupopo AirFlex Running Shoes"
                src="/images/ID3692_HM5.avif"
              />
            </div>
          </a>
          <form
            action="/mycart?index"
            class="_infoContainer_fa5548"
            data-discover="true"
            method="post"
          >
            <div
              class="_name_fa5548"
            >
              Jupopo AirFlex Running Shoes
            </div>
            <div
              class="_stars_fa5548"
            >
              <div
                class="_stars_container_fa5548"
              >
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548"
                  role="img"
                />
              </div>
              <div
                class="_ratings_fa5548"
              >
                (
                198
                )
              </div>
            </div>
            <div
              class="_price_fa5548"
            >
              $
              89.99
            </div>
            <button
              class="_addButton_fa5548"
            >
              Add to Cart
            </button>
            <input
              name="id"
              type="hidden"
              value="7"
            />
            <input
              name="name"
              type="hidden"
              value="Jupopo AirFlex Running Shoes"
            />
            <input
              name="style"
              type="hidden"
              value=""
            />
            <input
              name="price"
              type="hidden"
              value="89.99"
            />
            <input
              name="quantity"
              type="hidden"
              value="1"
            />
            <input
              name="picture"
              type="hidden"
              value="/images/ID3692_HM5.avif"
            />
            <input
              name="product"
              type="hidden"
              value="{"name":"Jupopo AirFlex Running Shoes","id":7,"category":"Men's Clothing","price":89.99,"ratings":198,"stars":4,"likes":431,"styles":[{"description":"","picture":"/images/ID3692_HM5.avif","isCurrentStyle":true}]}"
            />
          </form>
          <dialog
            role="dialog"
          >
            <img />
            <h2 />
            <span />
            <span />
            <span>
              Quantity: 
            </span>
            <button
              data-testid="g"
              type="button"
            >
              Ok
            </button>
          </dialog>
        </div>
        <div
          class="_viewedItem_fa5548"
        >
          <a
            data-discover="true"
            href="/product/8"
          >
            <div
              class="_imgContainer_fa5548"
            >
              <img
                alt="Velmora Elegance Handbag"
                src="/images/handbagblack.webp"
              />
            </div>
          </a>
          <form
            action="/mycart?index"
            class="_infoContainer_fa5548"
            data-discover="true"
            method="post"
          >
            <div
              class="_name_fa5548"
            >
              Velmora Elegance Handbag
            </div>
            <div
              class="_stars_fa5548"
            >
              <div
                class="_stars_container_fa5548"
              >
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
              </div>
              <div
                class="_ratings_fa5548"
              >
                (
                564
                )
              </div>
            </div>
            <div
              class="_price_fa5548"
            >
              $
              120
            </div>
            <button
              class="_addButton_fa5548"
            >
              Add to Cart
            </button>
            <input
              name="id"
              type="hidden"
              value="8"
            />
            <input
              name="name"
              type="hidden"
              value="Velmora Elegance Handbag"
            />
            <input
              name="style"
              type="hidden"
              value="Black"
            />
            <input
              name="price"
              type="hidden"
              value="120"
            />
            <input
              name="quantity"
              type="hidden"
              value="1"
            />
            <input
              name="picture"
              type="hidden"
              value="/images/handbagblack.webp"
            />
            <input
              name="product"
              type="hidden"
              value="{"name":"Velmora Elegance Handbag","id":8,"category":"Women's Clothing","price":120,"ratings":564,"stars":5,"likes":1201,"styles":[{"description":"Black","picture":"/images/handbagblack.webp","isCurrentStyle":true},{"description":"Beige","picture":"/images/1_536aef2b-2ba4-4bb5-98ec-6d6ac443cb43.webp","isCurrentStyle":false}]}"
            />
          </form>
          <dialog
            role="dialog"
          >
            <img />
            <h2 />
            <span />
            <span />
            <span>
              Quantity: 
            </span>
            <button
              data-testid="g"
              type="button"
            >
              Ok
            </button>
          </dialog>
        </div>
        <div
          class="_viewedItem_fa5548"
        >
          <a
            data-discover="true"
            href="/product/9"
          >
            <div
              class="_imgContainer_fa5548"
            >
              <img
                alt="Orvella Silver Twist Earrings"
                src="/images/twistearrings.webp"
              />
            </div>
          </a>
          <form
            action="/mycart?index"
            class="_infoContainer_fa5548"
            data-discover="true"
            method="post"
          >
            <div
              class="_name_fa5548"
            >
              Orvella Silver Twist Earrings
            </div>
            <div
              class="_stars_fa5548"
            >
              <div
                class="_stars_container_fa5548"
              >
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548"
                  role="img"
                />
              </div>
              <div
                class="_ratings_fa5548"
              >
                (
                310
                )
              </div>
            </div>
            <div
              class="_price_fa5548"
            >
              $
              59.99
            </div>
            <button
              class="_addButton_fa5548"
            >
              Add to Cart
            </button>
            <input
              name="id"
              type="hidden"
              value="9"
            />
            <input
              name="name"
              type="hidden"
              value="Orvella Silver Twist Earrings"
            />
            <input
              name="style"
              type="hidden"
              value=""
            />
            <input
              name="price"
              type="hidden"
              value="59.99"
            />
            <input
              name="quantity"
              type="hidden"
              value="1"
            />
            <input
              name="picture"
              type="hidden"
              value="/images/twistearrings.webp"
            />
            <input
              name="product"
              type="hidden"
              value="{"name":"Orvella Silver Twist Earrings","id":9,"category":"Jewelry","price":59.99,"ratings":310,"stars":4,"likes":720,"styles":[{"description":"","picture":"/images/twistearrings.webp","isCurrentStyle":true}]}"
            />
          </form>
          <dialog
            role="dialog"
          >
            <img />
            <h2 />
            <span />
            <span />
            <span>
              Quantity: 
            </span>
            <button
              data-testid="g"
              type="button"
            >
              Ok
            </button>
          </dialog>
        </div>
        <div
          class="_viewedItem_fa5548"
        >
          <a
            data-discover="true"
            href="/product/10"
          >
            <div
              class="_imgContainer_fa5548"
            >
              <img
                alt="AxionBeat Wireless Headphones"
                src="/images/headlessblue.webp"
              />
            </div>
          </a>
          <form
            action="/mycart?index"
            class="_infoContainer_fa5548"
            data-discover="true"
            method="post"
          >
            <div
              class="_name_fa5548"
            >
              AxionBeat Wireless Headphones
            </div>
            <div
              class="_stars_fa5548"
            >
              <div
                class="_stars_container_fa5548"
              >
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
                <span
                  aria-label="star"
                  class="_star_fa5548 _on_fa5548"
                  role="img"
                />
              </div>
              <div
                class="_ratings_fa5548"
              >
                (
                785
                )
              </div>
            </div>
            <div
              class="_price_fa5548"
            >
              $
              199.99
            </div>
            <button
              class="_addButton_fa5548"
            >
              Add to Cart
            </button>
            <input
              name="id"
              type="hidden"
              value="10"
            />
            <input
              name="name"
              type="hidden"
              value="AxionBeat Wireless Headphones"
            />
            <input
              name="style"
              type="hidden"
              value="Blue"
            />
            <input
              name="price"
              type="hidden"
              value="199.99"
            />
            <input
              name="quantity"
              type="hidden"
              value="1"
            />
            <input
              name="picture"
              type="hidden"
              value="/images/headlessblue.webp"
            />
            <input
              name="product"
              type="hidden"
              value="{"name":"AxionBeat Wireless Headphones","id":10,"category":"Electronics","price":199.99,"ratings":785,"stars":5,"likes":1532,"styles":[{"description":"Blue","picture":"/images/headlessblue.webp","isCurrentStyle":true},{"description":"White","picture":"/images/headlesswhite.webp","isCurrentStyle":false}]}"
            />
          </form>
          <dialog
            role="dialog"
          >
            <img />
            <h2 />
            <span />
            <span />
            <span>
              Quantity: 
            </span>
            <button
              data-testid="g"
              type="button"
            >
              Ok
            </button>
          </dialog>
        </div>
      </div>
    </div>
    <dialog>
      <form
        action="/mycart"
        data-discover="true"
        method="post"
      >
        <input
          name="id"
          type="hidden"
          value=""
        />
        <p>
          Remove the following item from your cart?
        </p>
        <p />
        <button
          aria-label="cancel"
          type="button"
        >
          Cancel
        </button>
      </form>
    </dialog>
  </div>
</div>
`;
